File: Anna/Source/Engine/OperatingSystems/Windows/Processes/ProcessCollector.fs
namespace Anna.Source.Engine.OperatingSystems.Windows.Processes
(*
    * In this class we fetch a process and store it in the target process passed by reference. The method of grabbing
    * processes and sorting them based on time since execution is as follows:
    * 1) Grab all available processes
    * 2) Sort them into two categories -- 'session0' (important system processes) and 'standard'.
    * It is worth noting here that we cannot access the 'time since execution' for 'session0' unless anathena is running
    * as admin. Trying to access the time results in errors and creates noticable overhead with a try/catch statement,
    * thus we have to sort them into the two formerly mentioned categories in advanced, and only fetch icons
    * for those in the 'standard' category.
    * 3) Sort the 'standard' list based on time since execution, and the 'session0' list based on processID
    * 4) Merge lists into one list, placing the 'standard' before the 'session0'.
    * 5) Loop over the 'standard' portion of the list, fetching icons.
    * Here it is also worth noting that there are issues trying to access an icon of a 64-bit process from a 32-bit
    * version of A. If we are 64-bit, we call a function that doesn't have to worry about this stuff. If we are
    * 32-bit, again try/catches again create too much overhead, so we use the function IsWow64Process to determine if
    * each process is compatable (also 32-bit), and if so THEN we can make a proper request.
    * 6) Update the target process in the static class TargetProcess
    * 
    * -------------------------------------------------------------------------------
    * 
    * Further implamentations:
    * - Icon fetching for session0 items (~3-5 have icons)
*)

open System
open System.Collections
open System.Diagnostics
open System.Drawing
open Anna.Source.Engine.OperatingSystems
open Anna.Source.Engine.OperatingSystems.Windows.Native

// Temporary structure for constructing process list
[<NoComparison>]
type IntermediateProcess ={ isSystemProcess: Boolean; externalProcess: Process }

type ProcessCollector() = 
    // Retrieves all running processes
    member this.GetProcesses() =
        Process.GetProcesses() |>
        Seq.map(fun (externalProcess) ->
        {
            isSystemProcess = this.IsProcessSystemProcess(externalProcess)
            externalProcess = externalProcess
        }) |>
        Seq.map(fun (filteredProcess) ->
        {
            processId= filteredProcess.externalProcess.Id
            processName= filteredProcess.externalProcess.ProcessName
            startTime = 
                try if filteredProcess.isSystemProcess then DateTime.MinValue else filteredProcess.externalProcess.StartTime
                with | _ -> DateTime.MinValue
            isSystemProcess= filteredProcess.isSystemProcess
            icon = this.GetIcon(filteredProcess.externalProcess, filteredProcess.isSystemProcess)
        }) |>
        Seq.sort

    // Fetches the icon associated with the provided process
    member this.IsProcessSystemProcess(externalProcess: Process) = 
        if (externalProcess.SessionId = 0 || externalProcess.BasePriority = 13) then true
        else
            try
                // Accessing this field will cause an access exception for system processes. This saves
                // time because handling the exception is faster than failing to fetch the icon later
                ignore externalProcess.PriorityBoostEnabled

                false
            with
                | _ -> true

    // Fetches the icon associated with the provided process
    member this.GetIcon(externalProcess: Process, isSystemProcess: Boolean) = 
        let noIcon: Icon = null
        // TODO: Check process bitness compatability
        if isSystemProcess then noIcon
        else
            try
                let iconHandle = NativeMethods.ExtractIcon(externalProcess.Handle, externalProcess.MainModule.FileName, 0)
                if (iconHandle = IntPtr.Zero) then noIcon else Icon.FromHandle(iconHandle)
            with
                | _ -> noIcon

File: Ana/Source/Engine/Processes/ProcessCollector.cs
namespace Ana.Source.Engine.Processes
{
    using OperatingSystems.Windows.Native;
    using System;
    using System.Collections.Generic;
    using System.Diagnostics;
    using System.Drawing;
    using System.Linq;

    /// <summary>
    /// A class responsible for collecting all running processes on the system
    /// </summary>
    public class ProcessCollector
    {
        /// <summary>
        /// Retrieves all running processes
        /// </summary>
        /// <returns>A collection of normalized processes</returns>
        public IEnumerable<NormalizedProcess> GetProcesses()
        {
            return Process.GetProcesses()
                .Select(externalProcess => new IntermediateProcess(this.IsProcessSystemProcess(externalProcess), externalProcess))
                .Select(intermediateProcess => new NormalizedProcess(
                        intermediateProcess.ExternalProcess.Id,
                        intermediateProcess.ExternalProcess.ProcessName,
                        intermediateProcess.ExternalProcess.StartTime,
                        intermediateProcess.IsSystemProcess,
                        this.GetIcon(intermediateProcess)))
                .OrderByDescending(normalizedProcess => normalizedProcess.processId);
        }

        /// <summary>
        /// Determines if the provided process is a system process
        /// </summary>
        /// <param name="externalProcess">The process to check</param>
        /// <returns>A value indicating whether or not the given process is a system process</returns>
        private Boolean IsProcessSystemProcess(Process externalProcess)
        {
            if (externalProcess.SessionId == 0 || externalProcess.BasePriority == 13)
            {
                return true;
            }

            try
            {
                if (externalProcess.PriorityBoostEnabled)
                {
                    // Accessing this field will cause an access exception for system processes. This saves
                    // time because handling the exception is faster than failing to fetch the icon later
                }

                return false;
            }
            catch
            {
                return true;
            }
        }

        /// <summary>
        /// Fetches the icon associated with the provided process
        /// </summary>
        /// <param name="intermediateProcess">An intermediate process structure</param>
        /// <returns>An Icon associated with the given process. Returns null if there is no icon</returns>
        private Icon GetIcon(IntermediateProcess intermediateProcess)
        {
            const Icon NoIcon = null;

            if (intermediateProcess.IsSystemProcess)
            {
                return NoIcon;
            }

            try
            {
                // TODO: This is a violation of the abstraction of native methods into just the OS adaptor. Either all process functions go into the OS Adapter,
                // or this portion must be moved into the Windows Adapter
                IntPtr iconHandle = NativeMethods.ExtractIcon(intermediateProcess.ExternalProcess.Handle, intermediateProcess.ExternalProcess.MainModule.FileName, 0);

                if (iconHandle == IntPtr.Zero)
                {
                    return NoIcon;
                }

                return Icon.FromHandle(iconHandle);
            }
            catch
            {
                return NoIcon;
            }
        }

        /// <summary>
        /// Temporary structure used in collecting all running processes
        /// </summary>
        private struct IntermediateProcess
        {
            /// <summary>
            /// Initializes a new instance of the <see cref="IntermediateProcess" /> struct
            /// </summary>
            /// <param name="isSystemProcess">Whether or not the process is a system process</param>
            /// <param name="externalProcess">The external process</param>
            public IntermediateProcess(Boolean isSystemProcess, Process externalProcess)
            {
                this.IsSystemProcess = isSystemProcess;
                this.ExternalProcess = externalProcess;
            }

            /// <summary>
            /// Gets a value indicating whether or not the process is a system process
            /// </summary>
            public Boolean IsSystemProcess { get; private set; }

            /// <summary>
            /// Gets the process associated with this intermediate structure
            /// </summary>
            public Process ExternalProcess { get; private set; }
        }
    }
    //// End class
}
//// End namespace